src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DataCollectionLauncher.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.DataCollection
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Text;
+
+    using Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.DataCollection.Interfaces;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+    using Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces;
+
+    /// <summary>
+    /// Abstract DataCollection Launcher provides functionality to handle process launch and exit events.
+    /// </summary>
+    public abstract class DataCollectionLauncher : IDataCollectionLauncher
+    {
+        protected IProcessHelper processHelper;
+
+        protected IMessageLogger messageLogger;
+
+        protected StringBuilder processStdError;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DataCollectionLauncher"/> class.
+        /// </summary>
+        /// <param name="processHelper">
+        /// The process helper.
+        /// </param>
+        /// <param name="messageLogger">
+        /// The message logger.
+        /// </param>
+        public DataCollectionLauncher(IProcessHelper processHelper, IMessageLogger messageLogger)
+        {
+            this.processHelper = processHelper;
+            this.messageLogger = messageLogger;
+            this.processStdError = new StringBuilder(this.ErrorLength, this.ErrorLength);
+        }
+
+        /// <inheritdoc />
+        public Process DataCollectorProcess { get; protected set; }
+
+        /// <summary>
+        /// Gets or sets the error length for data collector error stream.
+        /// </summary>
+        protected int ErrorLength { get; set; } = 4096;
+
+        /// <summary>
+        /// Gets callback on process exit
+        /// </summary>
+        protected Action<object> ExitCallBack => (process) =>
+        {
+            var exitCode = 0;
+            var processStdErrorStr = this.processStdError.ToString();
+
+            this.processHelper.TryGetExitCode(process, out exitCode);
+
+            if (exitCode != 0)
+            {
+                EqtTrace.Error("Data collector exited with error: '{0}'", processStdErrorStr);
+                this.messageLogger.SendMessage(TestMessageLevel.Error, processStdErrorStr);
```

Comment:
Or should we change the behavior of SendMessage() to not throw on empty message?